kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergWriterCompatV1MetadataValidatorAndUpdater.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.icebergcompat;
+
+import static io.delta.kernel.internal.tablefeatures.TableFeatures.*;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.TableConfig;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.tablefeatures.TableFeature;
+import io.delta.kernel.internal.tablefeatures.TableFeatures;
+import io.delta.kernel.internal.util.ColumnMapping;
+import io.delta.kernel.internal.util.SchemaUtils;
+import io.delta.kernel.internal.util.Tuple2;
+import io.delta.kernel.types.*;
+import java.util.*;
+import java.util.stream.Stream;
+
+/**
+ * Performs the validations and updates necessary to support the table feature IcebergWriterCompatV1
+ * when it is enabled by the table property "delta.enableIcebergWriterCompatV1".
+ *
+ * <p>Requires that the following table properties are set to the specified values. If they are set
+ * to an invalid value, throws an exception. If they are not set, enable them if possible.
+ *
+ * <ul>
+ *   <li>Requires ID column mapping mode (cannot be enabled on existing table).
+ *   <li>Requires icebergCompatV2 to be enabled.
+ * </ul>
+ *
+ * <p>Checks that required table features are enabled: icebergCompatWriterV1, icebergCompatV2,
+ * columnMapping
+ *
+ * <p>Checks the following:
+ *
+ * <ul>
+ *   <li>Checks that all table features supported in the table's protocol are in the allow-list of
+ *       table features. This simultaneously ensures that any unsupported features are not present
+ *       (e.g. CDF, variant type, etc).
+ *   <li>Checks that there are no fields with data type byte or short.
+ *   <li>Checks that the table feature `invariants` is not active in the table (i.e. there are no
+ *       invariants in the table schema). This is a special case where the incompatible feature
+ *       `invariants` is in the allow-list of features since it is included by default in the table
+ *       protocol for new tables. Since it is incompatible we must verify that it is inactive in the
+ *       table.
+ * </ul>
+ *
+ * TODO additional enforcements coming in (ie physicalName=fieldId)
+ */
+public class IcebergWriterCompatV1MetadataValidatorAndUpdater
```

Comment:
Could this class be a singleton? That is, private constructor, only have one actual INSTANCE method.

All of the instance APIs on this class just return private static final members.

Reply:
Yeah makes sense, same thing for `IcebergCompatV2MetadataValidatorAndUpdater` as well then